### PR TITLE
refactor checkEscape()

### PR DESCRIPTION
**Pull request by WalterBright**
This refactoring:
1. eliminates many unnecessary recursive instances of EscapeVisitor on the stack
2. eliminates wretched use of global gagging
3. adds explanatory comments

### CODE REVIEW

File: src/escape.c [hunk 1]
```diff
@@ -257,7 +289,7 @@ void checkEscapeRef(Scope *sc, Expression *e)
 
         void visit(PtrExp *e)
         {
-            checkEscape(sc, e->e1);
+            result |= checkEscape(sc, e->e1, gag);
```
Comment (andralex) [inline src/escape.c]: Do you want to still check this if result is already false?

**Comment by andralex** [comment]
rebase

File: src/escape.c [hunk 2]
```diff
@@ -92,7 +105,7 @@ void checkEscape(Scope *sc, Expression *e)
         {
             for (size_t i = 0; i < e->exps->dim; i++)
             {
-                checkEscape(sc, (*e->exps)[i]);
+                visit((*e->exps)[i]);
```
Comment (yebblies) [inline src/escape.c]: Did you mean `(*e->exps)[i]->accept()`?

File: src/escape.c (hunk 1 above)
Comment (WalterBright) [inline src/escape.c]: It is initialized to false and then stickily is set to true.

File: src/escape.c (hunk 2 above)
Comment (yebblies) [inline src/escape.c]: This is most likely why it's failing - no subexpressions are actually getting checked.

**Comment by WalterBright** [comment]
Ready to pull.

**Comment by yebblies** [comment]
It would be better if `result` was called something descriptive like `hadErrors`but it's not a big deal.

**Comment by yebblies** [comment]
Auto-merge toggled on

**Pull request merged**